### ISSUE TITLE
pgsphere regression test failures on ppc64el

### ISSUE THREAD

**Issue by ChristophBerg** [akorotkov/pgsphere#5]
Hi,
building pgsphere on ppc64el on PG 9.2 results in the regression diffs quoted below. All Debian and Ubuntu versions supported by apt.postgresql.org are affected; the diffs look similar for all distributions.
(I haven't yet tried running the regression tests against 9.3+; the package builds find with them, though, but likely the failures are the same. Also, the package builds and tests fine against 9.2..9.6 on amd64 and i386.)
Thanks!

```
00:01:38.658 ============== running regression test queries        ==============
00:01:38.878 test init                     ... ok
00:01:40.764 test tables                   ... ok
00:01:40.808 test points                   ... FAILED
00:01:40.830 test euler                    ... ok
00:01:40.873 test circle                   ... FAILED
00:01:40.900 test line                     ... ok
00:01:40.959 test ellipse                  ... ok
00:01:41.116 test poly                     ... ok
00:01:41.171 test path                     ... ok
00:01:41.264 test box                      ... ok
00:01:41.342 test index                    ... ok
00:01:41.419 test contains_ops             ... ok
00:01:41.482 test contains_ops_compat      ... ok
00:01:41.528 test bounding_box_gist        ... ok
00:01:41.583 test gnomo                    ... FAILED
00:01:41.583 
00:01:41.583 =======================
00:01:41.584  3 of 15 tests failed. 
00:01:41.584 =======================

00:01:41.616 2017-08-18 11:06:13.578 UTC [6563] postgres@contrib_regression ERROR:  spherepoly_in: more than two points needed at character 14
00:01:41.616 2017-08-18 11:06:13.578 UTC [6563] postgres@contrib_regression STATEMENT:  SELECT spoly '{(10d,0d),(10d,1d)}';
00:01:41.616 2017-08-18 11:06:13.578 UTC [6563] postgres@contrib_regression ERROR:  spherepoly_from_array: a line segment overlaps or polygon too large at character 14
00:01:41.616 2017-08-18 11:06:13.578 UTC [6563] postgres@contrib_regression STATEMENT:  SELECT spoly '{(0d,0d),(10d,10d),(0d,10d),(10d,0d)}';
00:01:41.616 Dropping cluster 9.2/regress ...
00:01:42.711 **** regression.diffs ****
00:01:42.712 *** /tmp/autopkgtest.vJaxRz/tree/expected/points.out	2017-07-28 15:35:48.000000000 +0000
00:01:42.712 --- /tmp/autopkgtest.vJaxRz/tree/results/points.out	2017-08-18 11:06:13.369260219 +0000
00:01:42.712 ***************
00:01:42.712 *** 110,124 ****
00:01:42.713   (1 row)
00:01:42.713   
00:01:42.713   SELECT '( 1h 2m 30s , -1d 2m 30s)'::spoint;
00:01:42.713 !             spoint             
00:01:42.713 ! -------------------------------
00:01:42.713 !  ( 15d 37m 30s , - 1d  2m 30s)
00:01:42.713   (1 row)
00:01:42.713   
00:01:42.713   SELECT '( 0h 2m 30s , -0d 2m 30s)'::spoint;
00:01:42.714 !             spoint             
00:01:42.714 ! -------------------------------
00:01:42.730 !  (  0d 37m 30s , - 0d  2m 30s)
00:01:42.730   (1 row)
00:01:42.730   
00:01:42.730   SELECT '( 0h 2m 30s , 90d 0m 0s)'::spoint;
00:01:42.730 --- 110,124 ----
00:01:42.730   (1 row)
00:01:42.730   
00:01:42.730   SELECT '( 1h 2m 30s , -1d 2m 30s)'::spoint;
00:01:42.736 !                spoint               
00:01:42.736 ! ------------------------------------
00:01:42.736 !  ( 15d 37m 30s , - 0d  0m -956250s)
00:01:42.736   (1 row)
00:01:42.736   
00:01:42.736   SELECT '( 0h 2m 30s , -0d 2m 30s)'::spoint;
00:01:42.736 !               spoint               
00:01:42.738 ! -----------------------------------
00:01:42.738 !  (  0d 37m 30s , - 0d  0m -38250s)
00:01:42.738   (1 row)
00:01:42.738   
00:01:42.738   SELECT '( 0h 2m 30s , 90d 0m 0s)'::spoint;
00:01:42.738 ***************
00:01:42.738 *** 128,136 ****
00:01:42.738   (1 row)
00:01:42.738   
00:01:42.739   SELECT '( 0h 2m 30s ,-90d 0m 0s)'::spoint;
00:01:42.739 !             spoint            
00:01:42.739 ! ------------------------------
00:01:42.739 !  (  0d 37m 30s , -90d  0m 0s)
00:01:42.739   (1 row)
00:01:42.739   
00:01:42.739   SELECT set_sphere_output( 'HMS' );
00:01:42.739 --- 128,136 ----
00:01:42.743   (1 row)
00:01:42.743   
00:01:42.743   SELECT '( 0h 2m 30s ,-90d 0m 0s)'::spoint;
00:01:42.743 !                 spoint                
00:01:42.743 ! --------------------------------------
00:01:42.743 !  (  0d 37m 30s , - 0d  0m -82620000s)
00:01:42.743   (1 row)
00:01:42.743   
00:01:42.743   SELECT set_sphere_output( 'HMS' );
00:01:42.744 ***************
00:01:42.744 *** 152,166 ****
00:01:42.744   (1 row)
00:01:42.744   
00:01:42.744   SELECT '( 1h 2m 30s , -1d 2m 30s)'::spoint;
00:01:42.744 !             spoint             
00:01:42.744 ! -------------------------------
00:01:42.744 !  (  1h  2m 30s , - 1d  2m 30s)
00:01:42.744   (1 row)
00:01:42.744   
00:01:42.744   SELECT '( 0h 2m 30s , -0d 2m 30s)'::spoint;
00:01:42.744 !             spoint             
00:01:42.744 ! -------------------------------
00:01:42.744 !  (  0h  2m 30s , - 0d  2m 30s)
00:01:42.744   (1 row)
00:01:42.745   
00:01:42.745   SELECT '( 0h 2m 30s , 90d 0m 0s)'::spoint;
00:01:42.745 --- 152,166 ----
00:01:42.745   (1 row)
00:01:42.745   
00:01:42.745   SELECT '( 1h 2m 30s , -1d 2m 30s)'::spoint;
00:01:42.745 !                spoint               
00:01:42.745 ! ------------------------------------
00:01:42.745 !  (  1h  2m 30s , - 0d  0m -956250s)
00:01:42.745   (1 row)
00:01:42.745   
00:01:42.745   SELECT '( 0h 2m 30s , -0d 2m 30s)'::spoint;
00:01:42.745 !               spoint               
00:01:42.745 ! -----------------------------------
00:01:42.745 !  (  0h  2m 30s , - 0d  0m -38250s)
00:01:42.745   (1 row)
00:01:42.746   
00:01:42.746   SELECT '( 0h 2m 30s , 90d 0m 0s)'::spoint;
00:01:42.746 ***************
00:01:42.746 *** 170,178 ****
00:01:42.746   (1 row)
00:01:42.746   
00:01:42.746   SELECT '( 0h 2m 30s ,-90d 0m 0s)'::spoint;
00:01:42.756 !             spoint            
00:01:42.756 ! ------------------------------
00:01:42.756 !  (  0h  2m 30s , -90d  0m 0s)
00:01:42.756   (1 row)
00:01:42.756   
00:01:42.756   SELECT set_sphere_output( 'RAD' );
00:01:42.756 --- 170,178 ----
00:01:42.756   (1 row)
00:01:42.757   
00:01:42.757   SELECT '( 0h 2m 30s ,-90d 0m 0s)'::spoint;
00:01:42.757 !                 spoint                
00:01:42.757 ! --------------------------------------
00:01:42.757 !  (  0h  2m 30s , - 0d  0m -82620000s)
00:01:42.757   (1 row)
00:01:42.757   
00:01:42.757   SELECT set_sphere_output( 'RAD' );
00:01:42.757 
00:01:42.761 ======================================================================
00:01:42.761 
00:01:42.761 *** /tmp/autopkgtest.vJaxRz/tree/expected/circle.out	2017-07-28 15:35:48.000000000 +0000
00:01:42.762 --- /tmp/autopkgtest.vJaxRz/tree/results/circle.out	2017-08-18 11:06:13.437261205 +0000
00:01:42.762 ***************
00:01:42.762 *** 215,223 ****
00:01:42.762   (1 row)
00:01:42.762   
00:01:42.762   SELECT scircle('(0d,-90d)'::spoint);
00:01:42.762 !                   scircle                   
00:01:42.762 ! --------------------------------------------
00:01:42.762 !  <(  0d  0m 0s , -90d  0m 0s) ,  0d  0m 0s>
00:01:42.763   (1 row)
00:01:42.763   
00:01:42.763   -- Operators
00:01:42.763 --- 215,223 ----
00:01:42.768   (1 row)
00:01:42.768   
00:01:42.768   SELECT scircle('(0d,-90d)'::spoint);
00:01:42.768 !                       scircle                       
00:01:42.768 ! ----------------------------------------------------
00:01:42.768 !  <(  0d  0m 0s , - 0d  0m -82620000s) ,  0d  0m 0s>
00:01:42.768   (1 row)
00:01:42.769   
00:01:42.769   -- Operators
00:01:42.769 
00:01:42.774 ======================================================================
00:01:42.775 
00:01:42.775 *** /tmp/autopkgtest.vJaxRz/tree/expected/gnomo.out	2017-07-28 15:35:48.000000000 +0000
00:01:42.775 --- /tmp/autopkgtest.vJaxRz/tree/results/gnomo.out	2017-08-18 11:06:14.141271403 +0000
00:01:42.775 ***************
00:01:42.775 *** 30,38 ****
00:01:42.775   (1 row)
00:01:42.775   
00:01:42.776   select gnomonic_proj(spoint '(1, -0.45)', spoint '(1, -0.45)');
00:01:42.776 !  gnomonic_proj 
00:01:42.776 ! ---------------
00:01:42.776 !  (0,0)
00:01:42.776   (1 row)
00:01:42.776   
00:01:42.776   select gnomonic_proj(spoint '(1, -0.445)', spoint '(1, -0.45)');
00:01:42.776 --- 30,38 ----
00:01:42.777   (1 row)
00:01:42.777   
00:01:42.777   select gnomonic_proj(spoint '(1, -0.45)', spoint '(1, -0.45)');
00:01:42.777 !      gnomonic_proj     
00:01:42.777 ! -----------------------
00:01:42.777 !  (0,6.54651790442e-18)
00:01:42.777   (1 row)
00:01:42.777   
00:01:42.777   select gnomonic_proj(spoint '(1, -0.445)', spoint '(1, -0.45)');
00:01:42.777 ***************
00:01:42.778 *** 84,92 ****
00:01:42.778   (1 row)
00:01:42.778   
00:01:42.778   select gnomonic_proj(spoint '(1, 0.05)', spoint '(1, 0.05)');
00:01:42.778 !  gnomonic_proj 
00:01:42.778 ! ---------------
00:01:42.778 !  (0,0)
00:01:42.778   (1 row)
00:01:42.778   
00:01:42.778   select gnomonic_proj(spoint '(1, 0.055)', spoint '(1, 0.05)');
00:01:42.778 --- 84,92 ----
00:01:42.778   (1 row)
00:01:42.779   
00:01:42.779   select gnomonic_proj(spoint '(1, 0.05)', spoint '(1, 0.05)');
00:01:42.779 !      gnomonic_proj     
00:01:42.779 ! -----------------------
00:01:42.779 !  (0,2.60434729872e-18)
00:01:42.779   (1 row)
00:01:42.779   
00:01:42.779   select gnomonic_proj(spoint '(1, 0.055)', spoint '(1, 0.05)');
00:01:42.779 ***************
00:01:42.779 *** 138,146 ****
00:01:42.779   (1 row)
00:01:42.779   
00:01:42.780   select gnomonic_proj(spoint '(1, 0.55)', spoint '(1, 0.55)');
00:01:42.780 !  gnomonic_proj 
00:01:42.780 ! ---------------
00:01:42.780 !  (0,0)
00:01:42.780   (1 row)
00:01:42.780   
00:01:42.780   select gnomonic_proj(spoint '(1, 0.555)', spoint '(1, 0.55)');
00:01:42.780 --- 138,146 ----
00:01:42.780   (1 row)
00:01:42.780   
00:01:42.780   select gnomonic_proj(spoint '(1, 0.55)', spoint '(1, 0.55)');
00:01:42.780 !      gnomonic_proj      
00:01:42.781 ! ------------------------
00:01:42.781 !  (0,-2.38158728306e-17)
00:01:42.781   (1 row)
00:01:42.781   
00:01:42.781   select gnomonic_proj(spoint '(1, 0.555)', spoint '(1, 0.55)');
00:01:42.781 ***************
00:01:42.781 *** 192,200 ****
00:01:42.781   (1 row)
00:01:42.781   
00:01:42.781   select gnomonic_proj(spoint '(3, -0.45)', spoint '(3, -0.45)');
00:01:42.781 !  gnomonic_proj 
00:01:42.781 ! ---------------
00:01:42.782 !  (0,0)
00:01:42.782   (1 row)
00:01:42.782   
00:01:42.782   select gnomonic_proj(spoint '(3, -0.445)', spoint '(3, -0.45)');
00:01:42.782 --- 192,200 ----
00:01:42.782   (1 row)
00:01:42.782   
00:01:42.782   select gnomonic_proj(spoint '(3, -0.45)', spoint '(3, -0.45)');
00:01:42.782 !      gnomonic_proj     
00:01:42.782 ! -----------------------
00:01:42.782 !  (0,6.54651790442e-18)
00:01:42.782   (1 row)
00:01:42.783   
00:01:42.783   select gnomonic_proj(spoint '(3, -0.445)', spoint '(3, -0.45)');
00:01:42.783 ***************
00:01:42.783 *** 246,254 ****
00:01:42.783   (1 row)
00:01:42.783   
00:01:42.783   select gnomonic_proj(spoint '(3, 0.05)', spoint '(3, 0.05)');
00:01:42.783 !  gnomonic_proj 
00:01:42.783 ! ---------------
00:01:42.783 !  (0,0)
00:01:42.783   (1 row)
00:01:42.783   
00:01:42.784   select gnomonic_proj(spoint '(3, 0.055)', spoint '(3, 0.05)');
00:01:42.784 --- 246,254 ----
00:01:42.784   (1 row)
00:01:42.784   
00:01:42.784   select gnomonic_proj(spoint '(3, 0.05)', spoint '(3, 0.05)');
00:01:42.784 !      gnomonic_proj     
00:01:42.784 ! -----------------------
00:01:42.790 !  (0,2.60434729872e-18)
00:01:42.797   (1 row)
00:01:42.797   
00:01:42.797   select gnomonic_proj(spoint '(3, 0.055)', spoint '(3, 0.05)');
00:01:42.797 ***************
00:01:42.797 *** 300,308 ****
00:01:42.798   (1 row)
00:01:42.798   
00:01:42.798   select gnomonic_proj(spoint '(3, 0.55)', spoint '(3, 0.55)');
00:01:42.798 !  gnomonic_proj 
00:01:42.798 ! ---------------
00:01:42.798 !  (0,0)
00:01:42.798   (1 row)
00:01:42.799   
00:01:42.799   select gnomonic_proj(spoint '(3, 0.555)', spoint '(3, 0.55)');
00:01:42.799 --- 300,308 ----
00:01:42.799   (1 row)
00:01:42.799   
00:01:42.799   select gnomonic_proj(spoint '(3, 0.55)', spoint '(3, 0.55)');
00:01:42.800 !      gnomonic_proj      
00:01:42.800 ! ------------------------
00:01:42.800 !  (0,-2.38158728306e-17)
00:01:42.800   (1 row)
00:01:42.800   
00:01:42.800   select gnomonic_proj(spoint '(3, 0.555)', spoint '(3, 0.55)');
00:01:42.801 ***************
00:01:42.801 *** 354,362 ****
00:01:42.801   (1 row)
00:01:42.801   
00:01:42.801   select gnomonic_proj(spoint '(5, -0.45)', spoint '(5, -0.45)');
00:01:42.802 !  gnomonic_proj 
00:01:42.802 ! ---------------
00:01:42.802 !  (0,0)
00:01:42.802   (1 row)
00:01:42.802   
00:01:42.802   select gnomonic_proj(spoint '(5, -0.445)', spoint '(5, -0.45)');
00:01:42.802 --- 354,362 ----
00:01:42.803   (1 row)
00:01:42.803   
00:01:42.803   select gnomonic_proj(spoint '(5, -0.45)', spoint '(5, -0.45)');
00:01:42.803 !      gnomonic_proj     
00:01:42.803 ! -----------------------
00:01:42.803 !  (0,6.54651790442e-18)
00:01:42.803   (1 row)
00:01:42.803   
00:01:42.803   select gnomonic_proj(spoint '(5, -0.445)', spoint '(5, -0.45)');
00:01:42.804 ***************
00:01:42.804 *** 408,416 ****
00:01:42.804   (1 row)
00:01:42.804   
00:01:42.804   select gnomonic_proj(spoint '(5, 0.05)', spoint '(5, 0.05)');
00:01:42.804 !  gnomonic_proj 
00:01:42.804 ! ---------------
00:01:42.804 !  (0,0)
00:01:42.804   (1 row)
00:01:42.804   
00:01:42.804   select gnomonic_proj(spoint '(5, 0.055)', spoint '(5, 0.05)');
00:01:42.804 --- 408,416 ----
00:01:42.805   (1 row)
00:01:42.805   
00:01:42.805   select gnomonic_proj(spoint '(5, 0.05)', spoint '(5, 0.05)');
00:01:42.805 !      gnomonic_proj     
00:01:42.805 ! -----------------------
00:01:42.805 !  (0,2.60434729872e-18)
00:01:42.805   (1 row)
00:01:42.805   
00:01:42.805   select gnomonic_proj(spoint '(5, 0.055)', spoint '(5, 0.05)');
00:01:42.805 ***************
00:01:42.805 *** 462,470 ****
00:01:42.806   (1 row)
00:01:42.806   
00:01:42.806   select gnomonic_proj(spoint '(5, 0.55)', spoint '(5, 0.55)');
00:01:42.806 !  gnomonic_proj 
00:01:42.806 ! ---------------
00:01:42.806 !  (0,0)
00:01:42.806   (1 row)
00:01:42.806   
00:01:42.806   select gnomonic_proj(spoint '(5, 0.555)', spoint '(5, 0.55)');
00:01:42.806 --- 462,470 ----
00:01:42.806   (1 row)
00:01:42.806   
00:01:42.806   select gnomonic_proj(spoint '(5, 0.55)', spoint '(5, 0.55)');
00:01:42.807 !      gnomonic_proj      
00:01:42.807 ! ------------------------
00:01:42.807 !  (0,-2.38158728306e-17)
00:01:42.808   (1 row)
00:01:42.808   
00:01:42.808   select gnomonic_proj(spoint '(5, 0.555)', spoint '(5, 0.55)');
00:01:42.808 
```

**Comment by mnullmei** [comment]
Well, the extant pgSphere code base is quite sensitive to floating-point rounding, especially in the regression tests (as was seen before with the double rounding from 80 bit FP to 64 bit FP that, regrettably, does occur with x87 FP instructions).
The "37m != 2m / -90d != -82620000s" bug probably is still something else, of course.

Today I was granted access on a ppc64le VM and I am looking into this right now.

**Comment by mnullmei** [comment]
OK, these problems are now fixed by a25e89db9eefffe76129207771df9c8fff48c3c5 and f25e203564d6d43633ef912b82e1ef9bc26ae4eb .

The Debian package is now updated as well ( https://anonscm.debian.org/git/pkg-postgresql/pgsphere.git/commit/?id=82e2bd0a49efbd9fad29d61c23ef24cebbca9993 ).
